timelock-agent/src/main/java/com/palantir/atlasdb/timelock/paxos/ClientPaxosResourceFactory.java

```diff
@@ -0,0 +1,264 @@
+import com.palantir.conjure.java.config.ssl.SslSocketFactories;
+import com.palantir.conjure.java.config.ssl.TrustContext;
+import com.palantir.paxos.PaxosAcceptor;
+import com.palantir.paxos.PaxosAcceptorNetworkClient;
+import com.palantir.paxos.PaxosLearner;
+import com.palantir.paxos.PaxosLearnerNetworkClient;
+import com.palantir.paxos.SingleLeaderAcceptorNetworkClient;
+import com.palantir.paxos.SingleLeaderLearnerNetworkClient;
+import com.palantir.timelock.config.TimeLockInstallConfiguration;
+import com.palantir.timelock.paxos.ClientAwarePaxosAcceptor;
+import com.palantir.timelock.paxos.ClientAwarePaxosAcceptorAdapter;
+import com.palantir.timelock.paxos.ClientAwarePaxosLearner;
+import com.palantir.timelock.paxos.ClientAwarePaxosLearnerAdapter;
+import com.palantir.timelock.paxos.PaxosRemotingUtils;
+
+public final class ClientPaxosResourceFactory {
+
+    private ClientPaxosResourceFactory() { }
+
+    public static ClientResources create(
+            MetricsManager metrics,
+            Path logDirectory,
+            TimeLockInstallConfiguration install,
+            ExecutorService sharedExecutor) {
+        PaxosComponents paxosComponents = new PaxosComponents(metrics.getTaggedRegistry(), "bound-store", logDirectory);
+        PaxosResource paxosResource = new PaxosResource(paxosComponents);
+        BatchPaxosAcceptorResource clientAcceptorResource =
+                new BatchPaxosAcceptorResource(new LocalBatchPaxosAcceptor(paxosComponents, new AcceptorCacheImpl()));
+        BatchPaxosLearnerResource clientLearnerResource = new BatchPaxosLearnerResource(paxosComponents);
+        UseCaseAwareBatchPaxosResource batchPaxosResource =
+                new UseCaseAwareBatchPaxosResource(clientAcceptorResource, clientLearnerResource);
+
+        int quorumSize = PaxosRemotingUtils.getQuorumSize(PaxosRemotingUtils.getClusterAddresses(install));
+
+        NetworkClientFactories factories;
+        if (install.paxos().clientPaxos().useBatchPaxos()) {
+            factories = batch(metrics.getRegistry(), install, quorumSize, sharedExecutor, batchPaxosResource);
+        } else {
+            factories = singleLeader(metrics.getRegistry(), install, quorumSize, sharedExecutor, paxosComponents);
+        }
+
+        return ImmutableClientResources.builder()
+                .quorumSize(quorumSize)
+                .components(paxosComponents)
+                .nonBatchedResource(paxosResource)
+                .batchedResource(batchPaxosResource)
+                .networkClientFactories(factories)
+                .addAllCloseables(factories.closeables())
+                .build();
+    }
+
+    @Value.Immutable
+    public interface ClientResources {
+        int quorumSize();
+        PaxosComponents components();
+        PaxosResource nonBatchedResource();
+        UseCaseAwareBatchPaxosResource batchedResource();
+        NetworkClientFactories networkClientFactories();
+        List<Closeable> closeables();
```

Comment:
nit: May be worth formalising if we guarantee anything about the order in which these are closed

Reply:
The order is super important, otherwise we end up blowing up timelock. I've currently opted to not do anything (akin to what we're already doing) but it's super sketchy.

I think for the most part, we need to walk along the creation path per client and note down anything that is creating executors, and either memoize them per client or pass them down via constructor otherwise we'll get screwed every time.

It's also something to consider for when we roll out the multi-leadership stuff, we should in theory spin down resources used by clients that we aren't going to be leaders of. But I suppose it's not so different from todays world.